core/src/mast/mod.rs

```diff
@@ -191,6 +202,61 @@ impl MastForest {
         self[node_id].set_after_exit(decorator_ids)
     }
 
+    /// Merges all `forests` into a new [`MastForest`].
+    ///
+    /// Merging two forests means combining all their constituent parts, i.e. [`MastNode`]s,
+    /// [`Decorator`]s and roots. During this process, any duplicate or
+    /// unreachable nodes are removed. Additionally, [`MastNodeId`]s of nodes as well as
```

Comment:
I brought this up before when @plafer looked into implementing a type of dead code elimination when building forests, and want to make sure we haven't unintentionally reintroduced the same issue:

How do you define whether a node is unreachable? With the presence of `DYN`, the only way to say for sure that a node is unreachable is one of two options:

* The node is not a procedure root (regardless of visibility, i.e. a procedure does not have to be exported to be dynamically invoked), _and_ it is not reachable by any other node. A very unlikely scenario, but nevertheless it is an edge case.
* The node is a procedure root with non-public visibility (i.e. not exported), and it's address (hash) is never captured via `procref` anywhere in its translation unit.

Procedure roots with public visibility are of course part of the public API of a library, so cannot ever be considered unreachable.

So as long as we are only removing nodes classified as unreachable by the above rules, that should be fine. However, AFAIK we haven't yet implemented the capturing of `procref` uses, so that leaves a very narrow set of nodes we can safely prune.

Reply:
Thanks for bringing this up!

I think it's easier to describe what is visited than what is not, due to how it is implemented:
The iterator that implements the DFS traversal iterates through all `MastForest::procedure_roots()`, and from each of those discovers the tree of that node ([this is the code](https://github.com/0xPolygonMiden/miden-vm/blob/8b6e66d81114713495c979a9b53d79ab81c5a285/core/src/mast/multi_forest_node_iterator.rs#L217
)). Hence all nodes reachable from procedure roots are visited. Other nodes are not visited. So since all nodes that are procedure roots and reachable from those nodes are visited and the others aren't I think your first point is covered, except maybe for the visibility part. One question I have is: Are the roots returned by `MastForest::procedure_roots()` only the public ones, or also the private ones? I'm guessing they're the public ones and there may be nodes in the forest that are procedures but aren't specified in `MastForest::roots`.

For the second point, what does `procref` translate to in the forest? An external node, but possibly one that is not in `MastForest::roots`? If so, then the node discovery would have to be changed.

To address everything at once, would the safest option perhaps be to iterate not over the `MastForest::procedure_roots` but over all nodes instead? We might copy dead nodes but for sure we wouldn't be removing nodes that are actually needed.

Reply:
> So since all nodes that are procedure roots and reachable from those nodes are visited and the others aren't I think your first point is covered

Yep! That sounds exactly right (of course, as I noted, it would be exceedingly rare for any nodes of a forest to be unreachable by this rule, since we construct the MastForest the same way, so such code would have been pruned when constructing the forest). That said, I think it is a good idea to mention this as you have in the docs.

> One question I have is: Are the roots returned by MastForest::procedure_roots() only the public ones, or also the private ones? I'm guessing they're the public ones and there may be nodes in the forest that are procedures but aren't specified in MastForest::roots.

/cc @plafer

I believe that originally produced _all_ of them, but I think that may have changed, and they are now only exported procedures, which is why I bring this up. Internally, we're supposed to be able to identify whether a given node is a non-exported procedure root though, so we should be able to recognize them when visiting the forest(s).

> For the second point, what does procref translate to in the forest? An external node, but possibly one that is not in MastForest::roots? If so, then the node discovery would have to be changed.

It gets rewritten to `push.HASH`, where `HASH` is the procedure root of the procedure it references. Later, the procedure would be called via `DYN`, with the procedure root on the stack (though maybe this is changing to be a memory address, rather than the procedure hash, but what I've described is how it has historically worked).

> To address everything at once, would the safest option perhaps be to iterate not over the MastForest::procedure_roots but over all nodes instead? We might copy dead nodes but for sure we wouldn't be removing nodes that are actually needed.

I think the key is to ensure that you visit all _roots_ regardless of whether they are part of the set of exported roots. IIRC that information is inside the MastForest (it must be, in order to implement `DYN`). As an aside, we should definitely ensure that a `MastForest` can be queried via its public API for _all_ procedure roots, not just exports, though we don't necessarily need to provide names for non-exported procedures like we do exports.

Dead code elimination will come later, when we introduce the analysis that we can use to implement that transformation, so I wouldn't worry about trying to do that here (it is better if we _don't_ actually).

Reply:
> As an aside, we should definitely ensure that a MastForest can be queried via its public API for all procedure roots, not just exports, though we don't necessarily need to provide names for non-exported procedures like we do exports.

Thinking on this further, I take back my point about not providing names, I think we also want to preserve names for non-exported procedures as well, so that we can pretty print stack traces in a debugger. Without that information, at best we'll only have source locations (file/line), but no names in a stack trace (unless it can be recovered by other means, but that shouldn't be relied upon IMO).

Reply:
> Are the roots returned by MastForest::procedure_roots() only the public ones, or also the private ones?

The `procedure_roots` contains public and private procedures (see [here](https://github.com/0xPolygonMiden/miden-vm/blob/62a113ee310390b08d3233807eb49a8ae89e5767/assembly/src/assembler/mod.rs#L521)) - we should definitely document this. So then the strategy is valid. 

> I brought this up before when @plafer looked into implementing a type of dead code elimination when building forests, and want to make sure we haven't unintentionally reintroduced the same issue:

You're right, and I keep forgetting about this issue, as a symptom that it's still not clear in my mind exactly what our strategy for `DYN` and dead code elimination is. For context, `DYN` is a bit tricky because technically we can call any value (even if not constructed with `procref`), and so if we're following the semantics to the letter, we should never be able to prune any node from a `MastForest`, because *technically*, it could be called with `DYN` (which the assembler has no way of knowing).

> With the presence of DYN, the only way to say for sure that a node is unreachable is one of two options:

This is probably the most reasonable contract: if you never `procref` a procedure, we're allowed to remove it from the `MastForest`. Is this what we want to go for? If so, we should clearly document it - if not only for my future self :).

So to recap, the current merging code doesn't perform any "unwanted dead code elimination", since it visits all nodes that are reachable by at least one procedure (whether private or public), and therefore *doesn't require any change related to this concern*. A future dead code elimination pass (AFAIU) would look at all local procedures that were not procref'd and remove them (as per our not-yet-documented contract that non-procref'd local procedures are considered dead).

Reply:
> The procedure_roots contains public and private procedures (see [here](https://github.com/0xPolygonMiden/miden-vm/blob/62a113ee310390b08d3233807eb49a8ae89e5767/assembly/src/assembler/mod.rs#L521)) - we should definitely document this. So then the strategy is valid.

Nice, so this PR is good on that front then, i.e. we aren't stripping things out we shouldn't be. Thanks for confirming that!

> For context, DYN is a bit tricky because technically we can call any value (even if not constructed with procref), and so if we're following the semantics to the letter, we should never be able to prune any node from a MastForest, because technically, it could be called with DYN (which the assembler has no way of knowing).

Agreed, technically this is true, however in prior discussions with @bobbinth, he has mentioned that he would like to restrict `DYN` to procedure roots, even though we haven't implemented such a restriction anywhere yet (and maybe we don't need to, really, you just end up in undefined behavior territory I suppose).

> This is probably the most reasonable contract: if you never procref a procedure, we're allowed to remove it from the MastForest. Is this what we want to go for? If so, we should clearly document it - if not only for my future self :).

Agreed, this is my operating assumption for whether or not we can consider a procedure dead. Personally, I don't think we should be doing dead code elimination of any kind at this level - leave it to the assembler or the compiler to do so, as they have more information anyway. That said, I don't have an objection to doing so, as long as it is done conservatively (i.e. we can _prove_ that a node is dead).

> A future dead code elimination pass (AFAIU) would look at all local procedures that were not procref'd and remove them

Essentially, yes. More formally: a procedure is proven dead if there are no static references to the procedure via some procedure invocation instruction (or `procref`/`external`), or such references only originate from procedures which are themselves marked dead.

Most likely, we'd want to use the call graph we construct in the assembler, ensuring it includes edges for `procref`, and then topologically ordering the nodes in that graph, and visiting from the leaves to the roots, removing dead leaves, and edges from dead non-leaf functions to their children (and revisiting the children when that happens to see if removing the edge made them dead as well). This could be done _before_ constructing the MastForest, though we would need to incorporate information from linked libraries, as well as determining what exported functions actually need to be preserved (i.e. exported procedures which are not part of the public interface of a library could be DCE'd if they are unused within the library).

Doing that work in the assembler makes the most sense to me, since it knows the full set of objects it is working with, and has most of the necessary metadata (if not all that is needed for this).

Anyway, I'm digressing, this PR is fine in terms of its dead code elimination effects, that's all that matters.